examples/download_progress/src/main.rs

```diff
@@ -1,144 +1,277 @@
+                    .on_press(Message::CancelDownload((*self.url).to_string())),
+                    TaskState::Finished => Button::new(
+                        &mut self.button,
+                        download_text("Downloaded"),
+                    ),
+                    TaskState::Error => {
+                        Button::new(&mut self.button, download_text("Errored"))
+                            .on_press(Message::StartDownload(
+                                (*self.url).to_string(),
+                            ))
+                    }
+                }
+                .width(Length::Units(85)),
+            )
+            .into()
+    }
+}
+
+impl Example {
+    async fn load_data() -> State {
+        let tasks: Vec<_> = (1..=10)
+            .map(|n| Task {
+                name: format!("File {:0>2}", n),
+                url: format!("{}?{}", FILE_URL, n),
+                ..Task::default()
+            })
+            .collect();
+        State {
+            tasks,
+            ..State::default()
+        }
+    }
 }
 
 impl Application for Example {
-    type Executor = executor::Default;
+    type Executor = iced::executor::Default;
     type Message = Message;
     type Flags = ();
 
     fn new(_flags: ()) -> (Example, Command<Message>) {
         (
-            Example::Idle {
-                button: button::State::new(),
-            },
-            Command::none(),
+            Example::Loading,
+            Command::perform(Example::load_data(), Message::Loaded),
         )
     }
 
     fn title(&self) -> String {
-        String::from("Download progress - Iced")
+        "Download Progress".to_string()
     }
 
-    fn update(&mut self, message: Message) -> Command<Message> {
-        match message {
-            Message::Download => match self {
-                Example::Idle { .. }
-                | Example::Finished { .. }
-                | Example::Errored { .. } => {
-                    *self = Example::Downloading { progress: 0.0 };
+    fn update(&mut self, message: Self::Message) -> Command<Self::Message> {
+        match self {
+            Example::Loading => {
+                if let Message::Loaded(state) = message {
+                    *self = Example::Loaded(state)
                 }
-                _ => {}
-            },
-            Message::DownloadProgressed(message) => match self {
-                Example::Downloading { progress } => match message {
-                    download::Progress::Started => {
-                        *progress = 0.0;
-                    }
-                    download::Progress::Advanced(percentage) => {
-                        *progress = percentage;
+            }
+            Example::Loaded(State {
+                tasks,
+                downloading_urls,
+                ..
+            }) => match message {
+                Message::StartDownload(url) => {
+                    if let Some(task) = tasks.iter_mut().find(|t| t.url == url)
```

Comment:
We could define `Message` as a struct with an `Url` and a `DownloadMessage`. This would allow us to easily find the index of the corresponding `Task` before matching on the particular `DownloadMessage`, removing duplication considerably as a consequence.